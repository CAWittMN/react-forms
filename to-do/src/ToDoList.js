import ToDo from "./ToDo";
import NewToDoForm from "./NewToDoForm";
import { useState } from "react";
import "./ToDoList.css";

const ToDoList = () => {
  const [todos, setTodos] = useState([]);

  const addToDo = (newToDo) => {
    setTodos([...todos, newToDo]);
  };

  const removeToDo = (id) => {
    setTodos(todos.filter((todo) => todo.id !== id));
  };

  const updateToDo = (id, updatedToDo) => {
    const updatedToDos = todos.map((todo) => {
      if (todo.id === id) {
        return { ...todo, text: updatedToDo };
      }
      return todo;
    });
    setTodos(updatedToDos);
  };

  const toDoComponents = todos.map((todo) => (
    <ToDo
      key={todo.id}
      id={todo.id}
      text={todo.text}
      removeToDo={removeToDo}
      updateToDo={updateToDo}
    />
  ));

  return (
    <div className="ToDoList">
      <NewToDoForm addToDo={addToDo} />
      {toDoComponents}
    </div>
  );
};

export default ToDoList;
