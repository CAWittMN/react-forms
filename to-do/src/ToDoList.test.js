import ToDoList from "./ToDoList";
import {
  render,
  fireEvent,
  getByDisplayValue,
  getByPlaceholderText,
} from "@testing-library/react";

it("renders without crashing", () => {
  render(<ToDoList />);
});

it("matches snapshot", () => {
  const { asFragment } = render(<ToDoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new todo", () => {
  const { getByLabelText, queryByText } = render(<ToDoList />);

  // no todos yet
  expect(queryByText("X")).not.toBeInTheDocument();

  // fill out the form
  const todoInput = getByLabelText("New To Do:");
  fireEvent.change(todoInput, { target: { value: "do laundry" } });

  // submit form
  const submitBtn = queryByText("Add To Do");
  fireEvent.click(submitBtn);

  // todo exists!
  expect(queryByText("X")).toBeInTheDocument();
});

it("can delete a todo", () => {
  const { getByLabelText, queryByText } = render(<ToDoList />);

  // fill out the form
  const todoInput = getByLabelText("New To Do:");
  fireEvent.change(todoInput, { target: { value: "do laundry" } });

  // submit form
  const submitBtn = queryByText("Add To Do");
  fireEvent.click(submitBtn);

  // todo exists!
  expect(queryByText("X")).toBeInTheDocument();

  // delete todo
  const deleteBtn = queryByText("X");
  fireEvent.click(deleteBtn);

  // todo is gone!
  expect(queryByText("X")).not.toBeInTheDocument();
});

it("can edit a todo", () => {
  const { getByLabelText, queryByText } = render(<ToDoList />);

  // fill out the form
  const todoInput = getByLabelText("New To Do:");
  fireEvent.change(todoInput, { target: { value: "do laundry" } });

  // submit form
  const submitBtn = queryByText("Add To Do");
  fireEvent.click(submitBtn);

  // todo exists!
  expect(queryByText("X")).toBeInTheDocument();

  // edit todo
  const editBtn = queryByText("Edit");
  fireEvent.click(editBtn);

  // edit form exists!
  expect(queryByText("Update")).toBeInTheDocument();

  // fill out the form
  const updatedTodoInput = getByLabelText("Edit:");
  fireEvent.change(updatedTodoInput, { target: { value: "do dishes" } });

  // submit form
  const updateBtn = queryByText("Update To Do");
  fireEvent.click(updateBtn);

  // todo is updated!
  expect(queryByText("do dishes")).toBeInTheDocument();
});
