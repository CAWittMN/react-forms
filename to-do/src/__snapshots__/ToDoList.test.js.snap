// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ToDoList"
  >
    <form
      class="NewToDoForm"
    >
      <label
        for="text"
      >
        New To Do:
      </label>
      <input
        id="text"
        name="text"
        type="text"
        value=""
      />
      <button>
        Add To Do
      </button>
    </form>
  </div>
</DocumentFragment>
`;
